Automattic/isolated-block-editor
# Add real time collaboration support

## Proposed changes

- Adds a `useYjs` React hook that adds real time collaborative editing features, powered by the [Yjs](https://github.com/yjs/yjs)-based [AsBlocks](https://github.com/youknowriad/asblocks). These features will only kick in when enabled and given a transport module via `settings.collab`.
- For local Storybook testing, we use a mock transport module that passes messages via the browser's Local Storage.
- Since "smart" undo (i.e. only undoing the user's own changes) is not implemented yet, undo is disabled while there are peers in the editor. This is to avoid data loss and user confusion.

![Diagram of useYjs](https://user-images.githubusercontent.com/555336/128883023-113f5aee-8532-41a4-af75-c610743af50f.png)

### TODO

- [x] commit peer edits to undo history
- [x] figure out reasonable logic to disable undo when it could delete peer edits
- [x] share presence and caret positions

### Punted

- Smart undo functionality that only undos the user's own changes. (I think we [need](https://docs.yjs.dev/api/undo-manager#example-specify-tracked-origins) to track change origins in transactions in AsBlocks. Related upstream issue: https://github.com/youknowriad/asblocks/issues/9)
- Making sure the caret labels are always positioned within the bounds of the editor.

## Testing instructions

1. `yarn install && yarn storybook` to open a Storybook story for the collaboration-enabled editor.
2. To see debug event messages, open devtools console and enter `localStorage.debug = 'iso-editor:collab'`.

Open two or more browser tabs with this same Storybook page, and see the edits sync across all the clients.
